Use the precedent of Rc::try_unwrap instead of the custom "take_or_clone_val"

ltratt opened the pull request:
**DON'T MERGE YET**

`take_or_clone_val` has two problems: 1) if you don't want to clone the value if the take fails, tough luck 2) it's a non-standard API.

This commit tentatively remodels the API along the lines of the standard library's `Rc::try_unwrap`. This has the virtue of familiarity and flexibility, but if you want to clone it's more of a pain. This:

```rust
  let x = cactus.take_or_clone_val().unwrap();
```

becomes:

```rust
  let x = cactus.try_unwrap().unwrap_or_else(|c| c.val().unwrap().clone())
```

I welcome thoughts as to whether this is a good idea orr not. I'm unsure.

ltratt committed:
Use the precedent of Rc::try_unwrap instead of the custom "take_or_clone_val".

take_or_clone_val has two problems: 1) if you don't want to clone the value
if the take fails, tough luck 2) it's a non-standard API.

This commit tentatively remodels the API along the lines of the standard
library's Rc::try_unwrap. This has the virtue of familiarity and flexibility,
but if you want to clone it's more of a pain. This:

  let x = cactus.take_or_clone_val().unwrap();

becomes:

  let x = cactus.try_unwrap().unwrap_or_else(|c| c.val().unwrap().clone())

vext01 commented:
Is it simpler to implement `take` on your cactus and leave the user to decide what to do if the clone fails?

vext01 commented on src/lib.rs:
```diff
@@ -179,24 +179,40 @@ impl<T> Cactus<T> {
-    /// is returned without cloning; if the node has children then the internal data is cloned
-    /// (hence why `T` must implement the `Clone` trait).
+    /// Try to consume this Cactus node and return its data. If the cactus node has no children,
+    /// this succeeds; if the cactus node has children, it fails, and returns the original
```
Wouldn't it be better to return something more obviously an error? Result?

ltratt replied:
This PR uses the API design from https://doc.rust-lang.org/std/rc/struct.Rc.html#method.try_unwrap as a guide.

ltratt commented:
Not sure I understand. `clone` can never fail. What can fail is the transfer of ownership. Or, in other words, `try_unwrap` basically says "if I succeed you can take over this memory without doing anything else" whereas `clone` has to `malloc` stuff and so on.

vext01 commented:
```
cactus.try_unwrap().unwrap_or_else(|c| c.val().unwrap().clone())
```

makes my head spin :(

vext01 commented:
Oops, I meant to say:

> Is it simpler to implement `take` on your cactus and leave the user to decide what to do if the `take` fails?

ltratt commented:
>Is it simpler to implement take on your cactus and leave the user to decide what to do if the take fails?

That's what this PR does (albeit calling it `try_unwrap` rather than `take`). If `try_unwrap` fails, you get the cactus back, and can then `clone` the val yourself.

vext01 commented:
Hrm, let me do some research. I hope we can find something better.

vext01 commented:
How about you have a function returns an enum which has two variants: `Taken` and `Cloned`, then `unwrap` extracts the inner value from whichever and always succeeds. If the user cares whether the value was cloned or not, they have the enum variant to inspect.

Downside: hand-rolled non-standard API.

vext01 commented:
Upside, you get something close to: `let x = cactus.take_or_clone_val().unwrap();`

ltratt commented:
Interesting, though, as you say, non-standard, idea. I think it's equivalent, though bulkier, to the current `take_or_clone_val`. One issue that `Rc::try_unwrap` solves is that you might want to say "if I can consume the node, fine; if I can't, I don't want to clone its contents". That's what this PR provides support for (but maybe it's a use case no-one cares about?).

vext01 commented:
In `Cloned` you would want to make the `clone()` lazy only if they unwrap.

Meh, not sure if I'm helping here...

ltratt commented:
Ah, I see what you mean.

Problem: if I don't want to clone, this will have consumed the cactus, and I can't get it back.

vext01 commented:
Ugh, you are right.

Alright, how about we implement both `take_or_clone_val` and `try_unwrap`?

Then we cover both use cases without imposing the extra hoops for people who don't mind if there's a clone or not.

ltratt commented:
Since it's a vaguely specialist use case, I think we should pick one of the methods and stick with it.

I've just converted some of my code that uses the cactus over to it, and I call this function precisely twice (despite having lots of Cactus code), so it's not ever likely to be a frequently called function.

vext01 commented:
Alright, well I don't know what's right really, so I'll stand by your decision.

I suppose long lines like the one you showed above are quite common in rust anyway...

Just a shame it makes my eyes burn :P

vext01 commented:
Ready to merge?

ltratt commented:
I'm still not sure if it's the right thing to do, but hey ho, it's code -- we can change it in the future if we change our mind :)

ltratt committed:
Bump version number to make new release.

Since we have changed the API, this means a major version increment.

ltratt commented:
I've added a new commit which will make a new major (gulp) release of cactus (since we've changed the API in this PR). If you're OK with that new commit, then I think this PR can be merged.

vext01 commented:
I probably would have gone with 0.2.0 myself, but versioning is somewhat arbitrary anyway.

OK by me. Any squashing before I merge?

ltratt commented:
If you read http://semver.org/ there's not really any wiggle room when you change the API in a backwards-incompatible way as this PR does. I'm not totally convinced that semantic versioning is solving a major problem, but it's probably best for us to stick to the standard.

I don't think this needs squashing.

vext01 commented:
Never heard of it!

Anyway, it's your project, so you choose the version number!
